Add support for matrix bot

avedor opened the pull request:
Hey @bjesus!

I figured it was about time to open the PR here for your review and feedback

Everything is tested and works with the exception of adding jobs for the scheduler. I haven't used this library before and am getting a value error `ValueError: Wrong number of fields; got 1, expected 5` and was wondering if you've run into something similar with the telegram bot.

Happy to hear your thoughts!

avedor committed:
Initial work to translate telegram to matrix

Remove unneeded functions, enable scheduler wip, bugfix no

Remove print statements, fix reply no, uncomment schedule

More selectively choose what to comment out

notify users about needed dependencies and configuration, fixes #1

Fix up how to do scheduling

Incorporate update to core re openai token

avedor commented:
I've gotten the scheduling to a semi-stable state. It only supports cron for now with a bit of validation, but should at least have feature parity with telegram

bjesus commented:
Hey @avedor ! sorry it took me so long to review this. it seems like the PR is doing two things - adding some Todoist backend _and_ adding the matrix bot.

I'm very happy to add Matrix support, but I'm a bit hesitant about implementing a different "backend" to Wren. I'm not entirely sure what's the use case and why not use one of many Todoist CLI tools instead? my first reaction would be that this is better off as another "plugin", just like Telegram or HTTP aren't affecting the `core` functionality of Wren.

avedor commented:
Hi @bjesus! Absolutely no issue! I hope all is well!

My apologies! It looks like I merged things incorrectly on my fork--I'd intended to keep the matrix branch completely separate. I've dropped the todoist related commits from the branch, so it should be back down to just the matrix commits.

Sorry again for any confusion!

bjesus commented:
Thank you! this looks much better. a few tiny comments before we merge this:

- can we change the setting `homeserver` to be called `matrix_homeserver` ?
- I think we need to also go over schedules at startup and register them, like we're doing at the telegram bot: https://github.com/bjesus/wren/blob/main/wren/telegram.py#L163-L169

bjesus commented:
also, let's use `croniter.is_valid()` to validate the cron format. I've just added that on the telegram side too: https://github.com/bjesus/wren/commit/f6e9111635e1f7e12737aaa97a85e558635adbd6#diff-5193d610d07d0454264253f854564b7521017a9b38bb14d31158612211c1d74dR131

bjesus commented:
lastly, let's add the matrix bot dependencies as optional deps, like the `http` and `telegram` deps: https://github.com/bjesus/wren/blob/main/setup.cfg#L23-L33

avedor committed:
Merge branch 'bjesus:main' into feature/add-matrix-bot

avedor committed:
Make deps optional; validate via croniter; register schedules at startup

Re-enable matrix cli arg; update scheduler

Switch verbiage from telegram to matrix

avedor commented:
hey @bjesus! i think i've gotten to everything you called out! i ran into a bit of trouble with the scheduler, though, as it wasn't separating the cron expression from the task name. i believe i've addressed that, but i wanted to call your attention to it because it does deviate from what you have in the telegram plugin

bjesus committed:
formatting

bjesus committed:
remove unneeded comments to match code style

bjesus commented:
looks good to me! just made some formatting changes to match the general style of the code. merging this now and will make a new release. thank you very much!

bjesus committed:
version 0.4.0

bjesus committed:
update readme with Matrix bot information
